### PR TITLE
Bump mimalloc to 2.0.2

### PR DESCRIPTION
**Pull request by ericriff**
Specify library name and version:  **mimalloc/2.0.2**

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/mimalloc/all/conanfile.py [hunk 1]
```diff
@@ -203,3 +217,8 @@ def package_info(self):
                 self.cpp_info.system_libs.extend(["psapi", "shell32", "user32", "bcrypt"])
             elif self.settings.os == "Linux":
                 self.cpp_info.system_libs.append("rt")
+
+        # Starting version 2.0 mimalloc installs the headers in a mimalloc-2.0 subfolder.
```
Comment (ericriff) [inline recipes/mimalloc/all/conanfile.py:221]: This is the most concerning part.
I'm exposing a path that shouldn't be exposed to make preserve 1.x compatibility on the test package.
1.x installs the headers loose at `include` whereas 2.x installs them at `include/mimalloc-2.0`. By exposing both paths the consumer is able to use the new path but also the older one.

Comment (madebr) [reply recipes/mimalloc/all/conanfile.py:221]: If the only reason is test_package compatibility, perhaps we need to rewrite its headers.
e.g.
```c
#if MIMALLOC_MAJOR_VERSION >= 2
#include "mimalloc-2.0/mimalloc.h"
#else
#include "mimalloc.h"
#endif
```

Comment (ericriff) [reply recipes/mimalloc/all/conanfile.py:221]: I take this back. 
The headers do get installed into `include/mimalloc-2.0`, but then they include each other with no `mimaloc-2.0` prefix.
E.g.
`include/mimalloc-2.0/mimalloc-override.h` has this line in it: `#include <mimalloc.h>`. So the only `includedir` to be exposed is `include/mimalloc-2.0`.

Comment (ericriff) [reply recipes/mimalloc/all/conanfile.py]: Starting version 1.7.0 they added this CMake option that disables all this subfolders nonsense. Enabling it makes the package layout "standard".
https://github.com/microsoft/mimalloc/blob/v1.7.0/CMakeLists.txt#L25

**Comment by uilianries** [comment]
Mimalloc 2.0.2 installs DLL file in lib/ folder, resulting in a runtime error when executing test_package.exe:


runtime_error


```
Copy mimalloc-redirect.dll to output directory
  -- Install configuration: "Release"
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/lib/mimalloc.lib
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/lib/mimalloc.dll
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/cmake/mimalloc.cmake
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/cmake/mimalloc-release.cmake
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/include/mimalloc.h
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/include/mimalloc-override.h
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/include/mimalloc-new-delete.h
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/cmake/mimalloc-config.cmake
  -- Installing: C:/Users/uilia/.conan/data/mimalloc/2.0.2/_/_/package/995e0f0b86a651012a3bfca00d60f35ae037db5e/cmake/mimalloc-config-version.cmake
```

You have to move that DLL file to bin/ folder.

**Comment by ericriff** [comment]
Thanks for looking into it. Should I move it with shutil or something like that or patch CMake so it uses `bin` as the `RUNTIME` folder?

**Comment by uilianries** [comment]
>  There is a patch for the older version that also patches this on CMake IIRC, if that's the case we might want to stick to that for consistency.

Yes, you can copy and use same patch. :+1:

**Comment by ericriff** [comment]
I don't think this is the right way of validating options
https://github.com/conan-io/conan-center-index/blob/master/recipes/mimalloc/all/conanfile.py#L57-L72

We're silently overriding the user options if they're not valid.
Should I replace all those option removals with `raise ConanInvalidConfiguration` ?

**Comment by madebr** [comment]
You can move these to `validate` and raise when the options are invalid.
But then, the package won't build with default options.

**Comment by ericriff** [comment]
Would changing the default values be a problem?

**Comment by madebr** [comment]
Normally, the default options are the same (or as close as possible) the options of the upstream project.
Perhaps it's ok to do only a check in `validate`...

**Comment by ericriff** [comment]
I haven't compared the default values of the options on the recipe with the sources. If the default values can't be built I would expect a mismatch between the recipe and the sources.

My plan was to make sure the default options on the recipe match the default options on the sources and then make the validations raise if we're picking up an invalid config. 
I don't like the idea of silently overriding what the user wants.

The default options of the sources should be a valid config.

**Comment by ericriff** [comment]
I don't understand what the `inject` option does. It seems to hide everything on `package_info()`?
Also the default value for the options do not match the default values on the sources. Should I fix that?

**Comment by madebr** [comment]
You can/may, but don't feel obliged to do that.
This can always be done in a later pr.

If you do, perhaps the default changed between versions?
In #2463, tarc had some issues getting it to work.
It can always have been an oversight.

**Review by madebr — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**